src/passes/param-utils.cpp

```diff
@@ -0,0 +1,177 @@
+                     const std::vector<CallRef*>& callRefs,
+                     Module* module,
+                     PassRunner* runner) {
+  assert(funcs.size() > 0);
+  auto* first = funcs[0];
+#ifndef NDEBUG
+  for (auto* func : funcs) {
+    assert(func->type == first->type);
+  }
+#endif
+
+  // Check if none of the calls has a param with side effects that we cannot
+  // remove (as if we can remove them, we will simply do that when we remove the
+  // parameter). Note: flattening the IR beforehand can help here.
+  bool callParamsAreValid =
+    std::none_of(calls.begin(), calls.end(), [&](Call* call) {
+      auto* operand = call->operands[index];
+      return EffectAnalyzer(runner->options, *module, operand)
+        .hasUnremovableSideEffects();
+    });
+  if (!callParamsAreValid) {
+    return false;
+  }
+
+  // The type must be valid for us to handle as a local (since we
+  // replace the parameter with a local).
+  // TODO: if there are no references at all, we can avoid creating a
+  //       local
+  bool typeIsValid = TypeUpdating::canHandleAsLocal(first->getLocalType(index));
+  if (!typeIsValid) {
+    return false;
+  }
+
+  // We can do it!
+
+  // Remove the parameter from the function. We must add a new local
+  // for uses of the parameter, but cannot make it use the same index
+  // (in general).
+  auto paramsType = first->getParams();
+  std::vector<Type> params(paramsType.begin(), paramsType.end());
+  auto type = params[index];
+  params.erase(params.begin() + index);
+  // TODO: parallelize some of these loops?
+  for (auto* func : funcs) {
+    func->setParams(Type(params));
+
+    // It's cumbersome to adjust local names - TODO don't clear them?
+    Builder::clearLocalNames(func);
+  }
+  std::vector<Index> newIndexes;
+  for (auto* func : funcs) {
+    newIndexes.push_back(Builder::addVar(func, type));
+  }
+  // Update local operations.
+  struct LocalUpdater : public PostWalker<LocalUpdater> {
+    Index removedIndex;
+    Index newIndex;
+    LocalUpdater(Function* func, Index removedIndex, Index newIndex)
+      : removedIndex(removedIndex), newIndex(newIndex) {
+      walk(func->body);
+    }
+    void visitLocalGet(LocalGet* curr) { updateIndex(curr->index); }
+    void visitLocalSet(LocalSet* curr) { updateIndex(curr->index); }
+    void updateIndex(Index& index) {
+      if (index == removedIndex) {
+        index = newIndex;
+      } else if (index > removedIndex) {
+        index--;
+      }
+    }
+  };
+  for (Index i = 0; i < funcs.size(); i++) {
+    auto* func = funcs[i];
+    if (!func->imported()) {
+      LocalUpdater(funcs[i], index, newIndexes[i]);
+      TypeUpdating::handleNonDefaultableLocals(func, *module);
+    }
+  }
+
+  // Remove the arguments from the calls.
+  for (auto* call : calls) {
+    call->operands.erase(call->operands.begin() + index);
+  }
+  for (auto* call : callRefs) {
+    call->operands.erase(call->operands.begin() + index);
+  }
```

Comment:
This doesn't seem to update `call_ref`s' target types. Is that OK?

Reply:
Ah, good point. Yes, this is intentional (the caller is expected to update types in all the places, its easier if a single place does it), but it should be documented. I'll open a PR.